File: app/src/main/java/com/example/earthwallet/ui/pages/wallet/ContactsAdapter.java
package com.example.earthwallet.ui.pages.wallet;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.earthwallet.R;
import com.example.earthwallet.wallet.services.ContactsManager;

import java.util.List;

/**
 * ContactsAdapter
 *
 * RecyclerView adapter for displaying contacts list
 */
public class ContactsAdapter extends RecyclerView.Adapter<ContactsAdapter.ContactViewHolder> {

    private final List<ContactsManager.Contact> contacts;
    private final OnContactClickListener clickListener;
    private final OnContactLongClickListener longClickListener;

    public interface OnContactClickListener {
        void onContactClick(ContactsManager.Contact contact);
    }

    public interface OnContactLongClickListener {
        void onContactLongClick(ContactsManager.Contact contact);
    }

    public ContactsAdapter(List<ContactsManager.Contact> contacts,
                          OnContactClickListener clickListener,
                          OnContactLongClickListener longClickListener) {
        this.contacts = contacts;
        this.clickListener = clickListener;
        this.longClickListener = longClickListener;
    }

    @NonNull
    @Override
    public ContactViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_contact, parent, false);
        return new ContactViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ContactViewHolder holder, int position) {
        ContactsManager.Contact contact = contacts.get(position);
        holder.bind(contact);
    }

    @Override
    public int getItemCount() {
        return contacts.size();
    }

    class ContactViewHolder extends RecyclerView.ViewHolder {
        private final TextView nameTextView;
        private final TextView addressTextView;

        ContactViewHolder(@NonNull View itemView) {
            super(itemView);
            nameTextView = itemView.findViewById(R.id.contactName);
            addressTextView = itemView.findViewById(R.id.contactAddress);
        }

        void bind(ContactsManager.Contact contact) {
            nameTextView.setText(contact.name);
            addressTextView.setText(formatAddress(contact.address));

            itemView.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onContactClick(contact);
                }
            });

            itemView.setOnLongClickListener(v -> {
                if (longClickListener != null) {
                    longClickListener.onContactLongClick(contact);
                }
                return true;
            });
        }

        private String formatAddress(String address) {
            if (address.length() > 20) {
                return address.substring(0, 14) + "..." + address.substring(address.length() - 6);
            }
            return address;
        }
    }
}

File: app/src/main/java/com/example/earthwallet/ui/pages/wallet/ContactsAdapter.kt
package network.erth.wallet.ui.pages.wallet

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import network.erth.wallet.R
import network.erth.wallet.wallet.services.ContactsManager

/**
 * ContactsAdapter
 *
 * RecyclerView adapter for displaying contacts list
 */
class ContactsAdapter(
    private val contacts: List<ContactsManager.Contact>,
    private val clickListener: OnContactClickListener?,
    private val longClickListener: OnContactLongClickListener?
) : RecyclerView.Adapter<ContactsAdapter.ContactViewHolder>() {

    interface OnContactClickListener {
        fun onContactClick(contact: ContactsManager.Contact)
    }

    interface OnContactLongClickListener {
        fun onContactLongClick(contact: ContactsManager.Contact)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_contact, parent, false)
        return ContactViewHolder(view)
    }

    override fun onBindViewHolder(holder: ContactViewHolder, position: Int) {
        val contact = contacts[position]
        holder.bind(contact)
    }

    override fun getItemCount(): Int = contacts.size

    inner class ContactViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val nameTextView: TextView = itemView.findViewById(R.id.contactName)
        private val addressTextView: TextView = itemView.findViewById(R.id.contactAddress)

        fun bind(contact: ContactsManager.Contact) {
            nameTextView.text = contact.name
            addressTextView.text = formatAddress(contact.address)

            itemView.setOnClickListener {
                clickListener?.onContactClick(contact)
            }

            itemView.setOnLongClickListener {
                longClickListener?.onContactLongClick(contact)
                true
            }
        }

        private fun formatAddress(address: String): String {
            return if (address.length > 20) {
                "${address.substring(0, 14)}...${address.substring(address.length - 6)}"
            } else {
                address
            }
        }
    }
}